fix(artifacts): Add support for github enterprise

jessesanford opened the pull request:
Signed-off-by: Jesse Sanford <jesse.sanford@autodesk.com>

Adds support for github enterprise. Currently the gui shows up broken because it doesn't understand URLs that don't contain "github". This will allow those as long as they are "mapped to github" in the list of scm accounts.

jessesanford commented on keel-core/src/test/kotlin/com/netflix/spinnaker/keel/services/ComparableLinksTests.kt:
```diff
@@ -287,11 +287,11 @@ class ComparableLinksTests : JUnit5Minutests {
           every { repository.artifactVersions(releaseArtifact, limit) } returns versions.toArtifactVersions(
             releaseArtifact,
-            "stash"
+            "https://stash"
```
Someone please review if it was critical that this work with commit urls that are protocol less. I think that was simply an oversight in writing the test.

dbyron-sf replied:
I don't know the code here to say one way or another, but putting this change in a separate commit would help clarify this....and make it easier to see that tests continue to pass before the other changes in the pr.

jessesanford replied:
Good point. I will restructure this.

jessesanford replied:
Now in explicit commit. Still calling attention to this though here so that it gets eyes :eyes:

nimakaviani replied:
previously, we didnt parse the link to extract the type so https in the source type didnt matter. better version for this change to go is probably here, where the link to the git artifact gets created: https://github.com/spinnaker/keel/blob/cf94218db7fc757c260dd5d866d478857dae0292/keel-core/src/test/kotlin/com/netflix/spinnaker/keel/services/ComparableLinksTests.kt#L157

jessesanford replied:
Are you suggesting that we update the mock generation code to convert "stash" -> "https://stash" rather than just use the string literal  "https://stash" below? What is the real world situation? Are these artifacts going to have protocol less sources? I like modifying the literal because the test reads more explicitly.

nimakaviani replied:
yeah either that or have `toArtifactVersions` and the `git` mock receive another optional parameter for `link`. Problem is, the second parameter for `toArtifactVersion` is the source type, and passing the protocol scheme as part of the source type makes it look weird.

jessesanford replied:
I understand what you mean and I agree, however aren't the original github tests passing protocol in via the second parameter? https://github.com/spinnaker/keel/blob/cf94218db7fc757c260dd5d866d478857dae0292/keel-core/src/test/kotlin/com/netflix/spinnaker/keel/services/ComparableLinksTests.kt#L620 should we clean all of these up for consistency in a separate PR?

nimakaviani replied:
yep, I agree with the inconsistencies, and it is a good idea to fix it while we are at it.

nimakaviani commented on keel-core/src/main/kotlin/com/netflix/spinnaker/keel/artifacts/ArtifactVersionLinks.kt:
```diff
@@ -35,14 +35,15 @@ class ArtifactVersionLinks(private val scmInfo: ScmInfo, private val cacheFactor
             "targetBranch=${olderGitMetadata.commitInfo?.sha}&sourceBranch=${newerGitMetadata.commitInfo?.sha}"
         }
-        "github" in commitLink -> {
+        "gitHub" in scmType -> {
```
can we make it case insensitive?

jessesanford replied:
Yes absolutley.

nimakaviani commented on keel-core/src/main/kotlin/com/netflix/spinnaker/keel/artifacts/ArtifactVersionLinks.kt:
```diff
@@ -51,22 +52,40 @@ class ArtifactVersionLinks(private val scmInfo: ScmInfo, private val cacheFactor
+      when {
+        "stash" in scmType ->
+          scmBaseURLs["stash"]
+        "gitHub" in scmType -> {
```
same here. problably better if `getScmType` converts the type to lower case to ensure consistency.

jessesanford replied:
I would rather perform the normalization of the string closer to the site of the check to avoid ambiguity. Check my latest commits.

nabuskey commented on keel-core/src/main/kotlin/com/netflix/spinnaker/keel/artifacts/ArtifactVersionLinks.kt:
```diff
@@ -51,22 +52,40 @@ class ArtifactVersionLinks(private val scmInfo: ScmInfo, private val cacheFactor
   }
 
+  fun getScmType(commitLink: String): String? {
+    val commitURL = URL(commitLink)
```
Do we need to worry about the potential exception being thrown here?

jessesanford replied:
Yes good call. I didn't like that. I fixed this by using getOrNull.

nabuskey replied:
I was talking about the`java.net.URL` constructor there. It throws `MalformedURLException`. So if given commit link is somehow malformed or empty, the exception will bubble up. `newerGitMetadata?.commitInfo?.link?` will check if it's not null, but not empty / blank string.

nabuskey commented on keel-core/src/main/kotlin/com/netflix/spinnaker/keel/artifacts/ArtifactVersionLinks.kt:
```diff
@@ -51,22 +52,40 @@ class ArtifactVersionLinks(private val scmInfo: ScmInfo, private val cacheFactor
+      commitURL.host == URL(baseUrl).host
+    }
+
+    return base.keys.toList()[0]
```
This could throw index out of boundary exception. If base is not found after filter, we should return null and log.

jessesanford replied:
didn't know about firstOrNull!, but I did use getOrNull let me know if you want me to change it.

nimakaviani commented on keel-core/src/main/kotlin/com/netflix/spinnaker/keel/artifacts/ArtifactVersionLinks.kt:
```diff
@@ -51,22 +53,41 @@ class ArtifactVersionLinks(private val scmInfo: ScmInfo, private val cacheFactor
+      commitURL.host == URL(baseUrl).host
+    }
+
+    return base.keys.toList().getOrNull(0)
```
probably slightly easier if `getScmType` return the lowercase type.

```suggestion
    return base.keys.toList().getOrNull(0)?.let { it.lowercase() }
```

jessesanford replied:
right but per my earlier comment, the intent is less explicit even if is more dry. A compromise could be renaming getScmType to getNormalizedScmType or getLowerScmType, I like the former in case there is more work to normalize these strings in the future.

nimakaviani replied:
yeah lets go with `getNormalizedScmType`. sounds good to me

nimakaviani commented on keel-core/src/main/kotlin/com/netflix/spinnaker/keel/artifacts/ArtifactVersionLinks.kt:
```diff
@@ -51,22 +53,41 @@ class ArtifactVersionLinks(private val scmInfo: ScmInfo, private val cacheFactor
-        "${url.protocol}://${url.host}"
+
+    return if (scmType != null) {
+      val scmTypeLower = scmType.lowercase()
```
```suggestion
```

nimakaviani commented:
build broken? 🤔

jessesanford commented:
I think we are good to go now. Your requested change can be marked resolved.

jessesanford commented:
Ok I removed the call to lowercase() because it's still marked as experimental (*shrug*) and instead am using toLowerCase (which is deprecated) but passes the tests. I also found an issue with the normalization that was causing the lookup to fail for github enterprise tests. So I added getScmType back in and dried up the code. Give a look please.

jessesanford commented:
Ok looks like builds are passing now. @nimakaviani @nabuskey

nimakaviani reviewed (approved):
LGTM
